systemd/cgroups.go

```diff
@@ -0,0 +1,263 @@
+func (c *CPUAcct) UsageSystemNanosecs() uint64 {
+	var nanoseconds uint64
+	for _, cpu := range c.CPUs {
+		nanoseconds += cpu.SystemNanosec
+	}
+	return nanoseconds
+}
+
+// UsageAllNanosecs returns total cpu consumption in nanoseconds, across all available cpu
+// cores, from the point that CPU accounting was enabled for this control group.
+func (c *CPUAcct) UsageAllNanosecs() uint64 {
+	var nanoseconds uint64
+	for _, cpu := range c.CPUs {
+		nanoseconds += cpu.SystemNanosec + cpu.UserNanosec
+	}
+	return nanoseconds
+}
+
+// ReadFileNoStat uses ioutil.ReadAll to read contents of entire file.
+// This is similar to ioutil.ReadFile but without the call to os.Stat, because
+// many files in /proc and /sys report incorrect file sizes (either 0 or 4096).
+// Reads a max file size of 512kB.  For files larger than this, a scanner
+// should be used.
+// COPIED FROM prometheus/procfs WHICH ALSO USES APACHE 2.0
+func ReadFileNoStat(filename string) ([]byte, error) {
+	const maxBufferSize = 1024 * 512
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	reader := io.LimitReader(f, maxBufferSize)
+	return ioutil.ReadAll(reader)
+}
+
+// NewCPUAcct will locate and read the kernel's cpu accounting info for
+// the provided systemd cgroup subpath.
+func NewCPUAcct(cgSubpath string) (*CPUAcct, error) {
+	var cpuUsage CPUAcct
+
+	cgPath, err := cgGetPath("cpu", cgSubpath, "cpuacct.usage_all")
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get cpu controller path")
+	}
+
+	// Example cpuacct.usage_all
+	// cpu user system
+	// 0 21165924 0
+	// 1 13334251 0
+	b, err := ReadFileNoStat(cgPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to read file %s", cgPath)
+	}
+
+	scanner := bufio.NewScanner(bytes.NewReader(b))
+	if ok := scanner.Scan(); !ok {
+		return nil, errors.Errorf("unable to scan file %s", cgPath)
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, errors.Wrapf(err, "unable to scan file %s", cgPath)
+	}
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return nil, errors.Wrapf(err, "unable to scan file %s", cgPath)
+		}
+		text := scanner.Text()
+		vals := strings.Split(text, " ")
+		if len(vals) != 3 {
+			return nil, errors.Errorf("unable to parse contents of file %s", cgPath)
+		}
+		cpu, err := strconv.ParseUint(vals[0], 10, 32)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to parse %s as uint32 (from %s)", vals[0], cgPath)
+		}
+		user, err := strconv.ParseUint(vals[1], 10, 64)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to parse %s as uint64 (from %s)", vals[1], cgPath)
+		}
+		sys, err := strconv.ParseUint(vals[2], 10, 64)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to parse %s as an in (from %s)", vals[2], cgPath)
+		}
+		onecpu := CPUUsage{
+			CPUId:         uint32(cpu),
+			UserNanosec:   user,
+			SystemNanosec: sys,
+		}
+		cpuUsage.CPUs = append(cpuUsage.CPUs, onecpu)
```

Comment:
Awesome for this to be merged, I just noticed it now, so sorry if  if this was asked before in the review.

Why even store array of `cpuUsages.CPUs`?

1) Why not aggregated information immediately?

2) Why not read aggregated data directly from `sysfs`? From `cpuacct.usage_sys` and `cpuacct.usage_user` files? It should provide exactly same information, and be faster, require way less parsing / aggregation, and be more scalable on high-CPU count systems. I see no point reading or using per-cpu data, or ever exporting per-cpu info, so it just doesn't make sense to read and handle this data in the first place IMHO.

Reply:
> Why not aggregated information immediately?

Mainly because I'm planning to upstream this - IMO there's little value in an exporter package long-term maintaining such parsing code. I've started doing some work in prometheus/procfs and I'm hoping to eventually have them maintain this code long-term. 

That said - I'll take a look at this in my next code push. I've noted some parts of procfs internally parse different files per method call (e.g. call a method to get per-cpu data, so they parse the per-cpu file. Call a method to get aggregate data, they parse the aggregate file). We could probably do something similar here, because you're right that this is needlessly wasteful. Most debian machines I use have about 300ish units, so with even 4 cores that is a fair bit of wasted effort+allocations. 

> Why not read aggregated data directly from sysfs? 

sysfs cannot expose data per systemd unit. We will want to be reading from cgroupfs if we want per-unit.

Reply:
FYI. By `sysfs` I meant `/sys/fs/cgroup`, aka `cgroupfs`. :)

I don't get your point about "mainly because I'm planning to upstream this". Reading directly from `cpuacct.usage_{user,sys}`, provides exactly same information, has no extra requirement, and is both faster, and easier to do with less code. I do not understand why you are doing it the hard / complex way.

Reply:
FYI. From performance point of view reading `cpuacct.usage_all` and doing immediete aggregation (in user-space) during parsing, might be slightly faster than reading aggregated data via `cpuacct.usage_user` and `cpuacct.usage_sys` as I suggested earlier. This is because in the first place (with suitably big read buffer) you only do 4 syscalls (open,read,read-zero,close), but in second you do 6-8 (for two separate files). There is no difference in atomicity (both `cpuacct.usage_user` and user-space aggregated `cpuacct.usage_all` can be slightly out of date, or not handled cpu hotplug fully (i.e. miss offlined cpus, and make the counte go backwards). `usage_all` might have minor cache locality boost too (i.e. because both sys and user times are read at the same time for each cpu, but in the `usage_user` and `usage_sys` they will be read separately, and thus could double the amount of cache traffic in most pathological conditions. But using an dynamic array (with `append`) in user-space will probably make this go opposite way in terms of performance. Aggregating in place from `cpuacct.usage_all` might be the way to go.